Multi-memory file handling

stefanrueger opened the pull request:
Fixes #1817

This PR handles r/w/v of multi-memory files that use a flat address space (as in .elf). Apparently Microchip use these

New functionality includes
 - Reading multi-memory *.hex* files to write single memories
 - Reading multi-memory *.srec* files to write single memories
 - Writing to a *list of memories* such as `flash,ee,fuses` from multi-memory files
 - Introducing `all` as an abbreviation for a list of all memories except submemories
 - Introducing `ALL` as an abbreviation for a list of all memories a part has
 - Introducing `etc` as the same as `all`: can use `sig,etc` which puts `signature` first in list (looks nicer than `sig,all`)
 - Enabling lists using an `\` (without) operator as in `all\bootrow`
 - Automatically *verifying when writing to multiple memories* via `-U [all | <list> | <list>\<list>]:w:file.hex:i`
 - Verifying a multi-memory file against a multi-memory file *later on* via `-U [<list>|all]:v:...`
 - Reading a list of memories to *write a multi-memory file* (se we can create our own multi-memory files)
 - Reading `all` memories to write a multi-memory file as *backup*
 - *Checking the signature of part* against the multi-memory file *signature section* to avoid overwriting the wrong part
 - New terminal command `backup <memlist> <file>[:format]`
 - New terminal command `restore <memlist> <file>[:format]`
 - New terminal command `verify <memlist> <file>[:format]`

In order to be able to *test* the new functionality, this PR also provides options `-xinit` for the dryrun programmer.

```
$ avrdude -c dryrun -xhelp
avrdude -c dryrun extended options:
  -xinit     Initialise memories with human-readable patterns (1, 2, 3)
  -xrandom   Initialise memories with random code/values (1, 3)
  -xseed=<n> Seed random number generator with <n>, default time(NULL)
  -xhelp     Show this help menu and exit
Notes:
  (1) -xinit and -xrandom randomly configure flash wrt boot/data/code length
  (2) Patterns can best be seen with fixed-width font on -U flash:r:-:I
  (3) Choose, eg, -xseed=1 for reproducible flash configuration and output
```

That is kinda fun, particularly, when you view the generated outputs using the `:I` format (.hex with comments). Try

```
$ avrdude -qq -c dryrun -p 16eb28 -xinit -xseed=7 -UALL:r:-:I  | more
```

So, feel free to do your worst to try to break the new, rewritten, update mechanism.

MCUdude commented:
Fantastic work @stefanrueger! Now this is an 8.0 feature for sure!

First test looks promising! Maybe `avrdude warning:` is a bit too much if a memory is missing in a multi-mem file? I think `avrdude:` would do.

```
$ ./avrdude -cpkobn_updi -pavr64du32 -Uall:w:/Users/hans/Downloads/hex-files/avr64du32-cnano-usb-cdc-to-usart-bridge-mplab-mcc.X_pro_1.0.0.hex 
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9621 (probably 64du32)

avrdude: processing -U all:w:/Users/hans/Downloads/hex-files/avr64du32-cnano-usb-cdc-to-usart-bridge-mplab-mcc.X_pro_1.0.0.hex:i
avrdude: reading 10034 bytes for multiple memories from input file /Users/hans/Downloads/hex-files/avr64du32-cnano-usb-cdc-to-usart-bridge-mplab-mcc.X_pro_1.0.0.hex
avrdude warning: /Users/hans/Downloads/hex-files/avr64du32-cnano-usb-cdc-to-usart-bridge-mplab-mcc.X_pro_1.0.0.hex has no data for eeprom, skipping ...
avrdude: 10022 bytes flash in 1 section [0, 0x2725]: 20 pages and 218 pad bytes
         writing 10022 bytes to flash ...
         writing | ################################################## | 100% 3.32 s 
         reading | ################################################## | 100% 1.86 s 
         10022 bytes of flash verified
avrdude: 12 bytes fuses in 1 section [0, 11]
         writing 12 bytes to fuses ..., 12 bytes written, 12 verified
avrdude warning: /Users/hans/Downloads/hex-files/avr64du32-cnano-usb-cdc-to-usart-bridge-mplab-mcc.X_pro_1.0.0.hex has no data for lock, skipping ...
avrdude warning: /Users/hans/Downloads/hex-files/avr64du32-cnano-usb-cdc-to-usart-bridge-mplab-mcc.X_pro_1.0.0.hex has no data for bootrow, skipping ...
avrdude warning: /Users/hans/Downloads/hex-files/avr64du32-cnano-usb-cdc-to-usart-bridge-mplab-mcc.X_pro_1.0.0.hex has no data for userrow/usersig, skipping ...

avrdude done.  Thank you.
```

Time is a bit limited at the moment, but I'll find some time to thoroughly go though the PR to understand all the new features this PR provides, and how they can be utilized.

MCUdude commented:
It maybe me that just doesn't get it, but I'm trying to write a backup multimem hex to an attiny2313 by ignoring the signature memory. How can this be done without using -F or a list of all memories, `flash,eeprom,lfuse,hfuse`?

```
$ ./avrdude -cdryrun -pattiny2313 -Uall\\sig:w:/Users/Hans/Downloads/t13_osccal_backup.hex:i
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e910a (probably t2313)

avrdude: processing -U all\sig:w:/Users/Hans/Downloads/t13_osccal_backup.hex:i
avrdude: reading 1096 bytes for multiple memories from input file /Users/Hans/Downloads/t13_osccal_backup.hex
avrdude error: signature of ATtiny2313 does not match file (ATtiny13, ATtiny13A)
        use -F to override this check

avrdude done.  Thank you.

```

stefanrueger commented:
Use ```` -F` . The check is deliberate. If the multi-memory file has the signature of the part it originated from it is *always* checked against the part it operates on.

It is rarely a good idea to transplant flash, fuses etc from a different MCU. Eeprom may be OK, but that's what -F is there for.

MCUdude commented:
> It is rarely a good idea to transplant flash, fuses etc from a different MCU. Eeprom may be OK, but that's what -F is there for.


It has happened several ties can't get the ATmega324P, and instead populate the 324A or 324PA instead. These are binary compatible, but in this case I'd have to use -F. _I_ think `all\sig` could be a would be a suitable option here.

dl8dtl commented:
Since I'm just testing an AVR16EB14 – the different pincount variants there also have different signatures, yet they are fully software compatible.

I'd prefer to not use backslashes. Why not `all-sig` or `all!sig`?

Amazing commit trail. ;-)

stefanrueger commented:
> AVR16EB14 – the different pincount variants there also have different signatures, yet they are fully software compatible

No (it's complicated, read on).

But yes, signature is a weak proxy for software-compatibility. More often than I thought are parts with different signature software-compatible (which I define as same flash size, same interrupt table, same configuration options and same register file), see table. And of course, reflashing board X with software from board Y can spell disaster even when both boards have the same MCU.

| Part | SW compat  | SW compat | SW compat |
| :--| :--| :--| :--|
|ATmega1284|ATmega1284P|||
|ATmega1284RFR2|ATmega128RFR2|||
|ATmega164A|ATmega164P=ATmega164PA|||
|ATmega165A|ATmega165P=ATmega165PA|||
|ATmega168A|ATmega168P=ATmega168PA|||
|ATmega169A|ATmega169P=ATmega169PA|||
|ATmega2564RFR2|ATmega256RFR2|||
|ATmega324A|ATmega324P|ATmega324PA||
|ATmega3250=ATmega3250A|ATmega3250P=ATmega3250PA|||
|ATmega325=ATmega325A|ATmega325P=ATmega325PA|||
|ATmega3290=ATmega3290A|ATmega3290P|||
|ATmega329A=ATmega329|ATmega329PA|||
|ATmega48A|ATmega48PA|||
|ATmega48|ATmega48P|||
|ATmega644A|ATmega644P=ATmega644PA|||
|ATmega644RFR2|ATmega64RFR2|||
|ATmega6450=ATmega6450A|ATmega6450P|||
|ATmega645=ATmega645A|ATmega645P|||
|ATmega6490A=ATmega6490|ATmega6490P|||
|ATmega649=ATmega649A|ATmega649P|||
|ATmega88=ATmega88A|ATmega88P=ATmega88PA|||
|ATtiny102|ATtiny104|||
|ATtiny1604|ATtiny1606|ATtiny1607||
|ATtiny1616|ATtiny1617|||
|ATtiny1626|ATtiny1627|||
|ATtiny3216|ATtiny3217|||
|ATtiny3226|ATtiny3227|||
|ATtiny416|ATtiny417|||
|ATtiny426|ATtiny427|||
|ATtiny804|ATtiny806|ATtiny807||
|ATtiny816|ATtiny817|||
|ATtiny826|ATtiny827|||
|AVR16DD14|AVR16DD20|AVR16DD28|AVR16DD32|
|AVR16DU20|AVR16DU28|AVR16DU32||
|AVR16EA28|AVR16EA32|||
|ATxmega128A1|ATxmega128A1revD|||
|AVR16EB20|AVR16EB28|AVR16EB32||
|AVR32DD14|AVR32DD20|AVR32DD28|AVR32DD32|
|AVR32DU20|AVR32DU28|AVR32DU32||
|AVR32EA28|AVR32EA32|||
|AVR64DD14|AVR64DD20|AVR64DD28|AVR64DD32|
|AVR64DU28|AVR64DU32|||
|AVR64EA28|AVR64EA32|||

Parts in different columns have different signature. part1=part2 means both have the same signature.

@dl8dtl According to AVRDUDE's understanding above the 16EB14 is not software-compatible with the 16EB20. And indeed,

```
$ avrdude -qqp 16eb14 -cdryrun -T "regf -a" >/tmp/1
$ avrdude -qqp 16eb20 -cdryrun -T "regf -a" | diff - /tmp/1
227,228d226
< I/O 0x05e8:       0x00 portmux.tcbroutea
< I/O 0x05ea:       0x00 portmux.acroutea
```

The 16EB20 seems to have two more registers. Maybe a sloppy ATDF? Or an ATDF generator that recognises that the 16EB14  has neither pin accessible that could be configured with that register? Or a mischievous developer who really has fun imagining how programmers struggle with errors thown by missing registers for the same MCU-family? ... who knows

Fun fact 2: avrdude.conf claims that `ATxmega128A1` and `ATxmega128A1revD` have different signature. I found no source for that. Maybe chuck `ATxmega128A1revD` out of `avrdude.conf` unless someone can conform this is is correct.

What should we do? Nothing? Remove the signature test? Only complain if the two signatures are not software-compatible? The test is simple (are the avrintel register/interrupt/config table pointers the same and do they have the same flash) but made more complicated by Atmel/Microchip's decision to ship different (and not software-compatible) parts with the same signature.

stefanrueger commented:
> Why not all-sig or all!sig?

Good question. I am not 100% happy with the way the memory list is parsed but for other reasons: the without operation can only be used once and has lower precedence than the list comma, ie, `flash,etc\signature,bootrow` is understood as the list `flash,etc` (basically `all` but in different order) without any in the list `signature,bootrow`.

Some shells would expand `!sig` to the last command that starts with `sig`, so needs quoting or escaping `\!` (as does `\` but at least that's easier to type and has the meaning of without in set theory).

`all-sig,bootrow` is weird for programmers as almost always `,` has lower precedence than `-` but it's the other way round here. 

Could do `-` and `+`, give them the same prio and parse from left to right (left associative). More expressive, easy to explain, does not confound programmers but `flash+eeprom+etc-sigrow+signature` looks decidedly ugly as list.

What do others think?

MCUdude commented:
> The 16EB20 seems to have two more registers

Not sure if that's sloppy ATDF work from Microchip's side or not, but I'd think that I'd compile code for the 16EB14 one, and that should work on a 16EB20 chip. Not necessarily the other way around, since chips with more pins usually also have more peripherals.

> What should we do? Nothing? Remove the signature test?

The signature test is fine. It's actually very useful! But I think it would be nice to be able to ignore it using `all\sig`, like with other memories. It's incredibly easy to write a hex file to the wrong microcontroller for non-multimem files, so we already have an extra layer of protection. So I think `all\sig` to bypass the signature check for the file is a good solution.

>Could do - and +, give them the same prio and parse from left to right (left associative). More expressive, easy to explain, does not confound programmers but flash+eeprom+etc-sigrow+signature looks decidedly ugly as list.
What do others think?

How about keeping the commas and only use `-` for memories to leave out? Like so?  
`-Uflash:w:ALL,-sig,-eeprom:w:file.hex` and `-Uflash,eeprom,fuses:w:file.hex`?

stefanrueger commented:
> More often than I thought are parts with different signature software-compatible

Made me wonder why. And, by and large, it's pin count (except for a handful of P/PA or similar cases). But I'd argue something is fishy when a program written for a larger package is put on a smaller package and vice versa (even though technically they are in some sense software-compatible they are not hardware-compatible).

> commas and only use - for memories to leave out

That's a vvv neat idea. I like that. Will probably implement `all,/sig,/eeprom` for not (as in signals). Have done so elsewhere.

MCUdude commented:
> But I'd argue something is fishy when a program written for a larger package is put on a smaller package and vice versa (even though technically they are in some sense software-compatible they are not hardware-compatible).

Yes, I wouldn't do that, but it can be done. However, sharing the same binary file between ATmega324A, P and PA isn't an issue at all, and something we do regularly when batch programming at work. Sometimes the 324P is unobtainable while the PA is readily available. so `-UALL,-sig` is something I'd definitely find useful for batch programming.

dl8dtl commented:
> The 16EB20 seems to have two more registers. Maybe a sloppy ATDF? Or an ATDF generator that recognises that the 16EB14 has neither pin accessible that could be configured with that register?

Maybe the latter.

I don't see PORTMUX.ACROUTE in the datasheet at all, though the IO header file in AVR-LibC has it. I see PORTMUX.TCBROUTE in the datasheet (with no mention on a dependency of the pincount), but indeed, any of the pins mentioned there aren't available on the low-pincount part.

My guess is that all of these MCUs use the same die anyway, just different bond patterns. The signature is OTP-fused into the parts on the IC tester anyway, so it's easy to flash different signatures there.

Unlike with different memory configurations, the peripherals are the same between all pinning variants.

> as does `\`

Well, yes, that's why I wouldn't prefer the backslash. In addition, at least to me at has no "mental relationship" to the action "subtract (or lessen)".

dl8dtl commented:
> I don't see PORTMUX.ACROUTE in the datasheet at all, though the IO header file in AVR-LibC has it.

Ah OK:
```
/* Analog Comparator 0 Output select */
typedef enum PORTMUX_AC0_enum
{
    PORTMUX_AC0_DEFAULT_gc = (0x00<<0)  /* OUT: PA7 */
} PORTMUX_AC0_t;

/* Analog Comparator 1 Output select */
typedef enum PORTMUX_AC1_enum
{
    PORTMUX_AC1_DEFAULT_gc = (0x00<<1)  /* OUT: PA7 */
} PORTMUX_AC1_t;
```
There's nothing to actually chose from, only the default. :-)

stefanrueger commented:
> `all\sig` to bypass the signature check

But what do you do if someone only wants to extract fuses and flash from a multi-memory file? They might extract that from the *wrong* file!  The signature check for multi-memory writes is triggered as soon as the *file* has a signature in it, not when signature appears in the list. `fuses,flash,/sig` evaluates to the same as `fuses,flash`.

```
$ avrdude -qq -c dryrun -p 16eb28 -xinit -xseed=7 -UALL:r:avr16eb28.bak:I
$  avrdude -qqp atmega4808 -cdryrun -U fuses,flash:w:avr16eb28.bak
avrdude error: signature of ATmega4808 does not match file (AVR16EB28)
        use -F to override this check
```

(I just realise that single-memory writes from multi-memory-files do not check the signature. Should be done.)

stefanrueger commented:
>  ATmega324A, P and PA isn't an issue at all

Yes, I understand. Funnily enough, AVRDUDE *can* figure out that these are SW-compatible, *but* it cannot figure out whether parts with different signature that are SW-compatible are *hardware*-incompatible (different pin counts). That's the difficulty.

> no "mental relationship" to the action "subtract (or lessen)".

It's standard [notation](https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement). But then I've had set theory first thing in secondary school (and you can tell I liked it!).

dl8dtl commented:
> > no "mental relationship" to the action "subtract (or lessen)".
> 
> It's standard [notation](https://en.wikipedia.org/wiki/Complement_(set_theory)#Relative_complement).

Hmm, I did have a bit of set theory, but don't remember that one … My proposal for the minus sign was as it is easy to type – backslash, and (as you noticed) exclamation marks need to be escaped from (Unix) shells, minus doesn't.

stefanrueger commented:
Seems like @MCUdude's (modified) idea might be winning: `ALL,/signa,/ee` (evaluated from left to right, '/' means but not)

mcuee commented:
Nice feature.

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1828.exe -C .\avrdude_pr1828.conf -qq -c dryrun -xhelp
avrdude_pr1828 -c dryrun extended options:
  -xinit     Initialise memories with human-readable patterns (1, 2, 3)
  -xrandom   Initialise memories with random code/values (1, 3)
  -xseed=<n> Seed random number generator with <n>, default time(NULL)
  -xhelp     Show this help menu and exit
Notes:
  (1) -xinit and -xrandom randomly configure flash wrt boot/data/code length
  (2) Patterns can best be seen with fixed-width font on -U flash:r:-:I
  (3) Choose, eg, -xseed=1 for reproducible flash configuration and output

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1828.exe -C .\avrdude_pr1828.conf -qq -c dryrun -p m328p -xinit -xseed=7 -UALL:r:-:I
:02000004008179
:2000000054686520717569636B2062726F776E20666F78206A756D7073206F76657220740E // 810000> The quick brown fox jumps over t eeprom
:200020006865206C617A7920646F672E2054686520717569636B2062726F776E20666F7858 // 810020> he lazy dog. The quick brown fox
:20004000206A756D7073206F76657220746865206C617A7920646F672E2054686520717565 // 810040>  jumps over the lazy dog. The qu
:2000600069636B2062726F776E20666F78206A756D7073206F76657220746865206C617AA1 // 810060> ick brown fox jumps over the laz
:200080007920646F672E2054686520717569636B2062726F776E20666F78206A756D7073DD // 810080> y dog. The quick brown fox jumps
:2000A000206F76657220746865206C617A7920646F672E2054686520717569636B20627229 // 8100a0>  over the lazy dog. The quick br
:2000C0006F776E20666F78206A756D7073206F76657220746865206C617A7920646F672E6B // 8100c0> own fox jumps over the lazy dog.
:2000E0002054686520717569636B2062726F776E20666F78206A756D7073206F7665722082 // 8100e0>  The quick brown fox jumps over
:20010000746865206C617A7920646F672E2054686520717569636B2062726F776E20666F7B // 810100> the lazy dog. The quick brown fo
:2001200078206A756D7073206F76657220746865206C617A7920646F672E20546865207181 // 810120> x jumps over the lazy dog. The q
:200140007569636B2062726F776E20666F78206A756D7073206F76657220746865206C61C5 // 810140> uick brown fox jumps over the la
:200160007A7920646F672E20FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFC // 810160> zy dog. ........................
......
:207FE000204020202020204020202020404020204020402020404020202020402020202061 // 07fe0>  @     @    @@  @ @  @@    @
:02000004008278
:01000000629D                                                               // 820000> 0x62 lfuse
:02000004008278
:01000100DD21                                                               // 820001> 0xdd hfuse
:02000004008278
:01000200FFFE                                                               // 820002> 0xff efuse
:02000004008377
:01000000FF00                                                               // 830000> 0xff lock
:02000004008476
:030000001E950F3B                                                           // 840000> 0x1e 0x95 0x0f signature (ATmega328P, ATA6614Q, LGT8F328P)
:02000004008476
:0100030055A7                                                               // 840003> 0x55 calibration
:00000001FF
```

mcuee commented:
I like the "ALL" option as well.

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude_pr1828.exe -C .\avrdude_pr1828.conf -qq -c usbasp -p m168p -UALL:r:readall_m168p.hex:I

 MINGW64 /c/work/avr/avrdude_test/avrdude_bin
$ head readall_m168p.hex
:02000004008179
:20000000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF00 // 810000> ................................ eeprom
:20002000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFE0 // 810020> ................................
:20004000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFC0 // 810040> ................................
:20006000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFA0 // 810060> ................................
:20008000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF80 // 810080> ................................
:2000A000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF60 // 8100a0> ................................
:2000C000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF40 // 8100c0> ................................
:2000E000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF20 // 8100e0> ................................
:20010000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF // 810100> ................................

$ tail readall_m168p.hex
:01000100DD21                                                               // 820001> 0xdd hfuse
:02000004008278
:01000200FC01                                                               // 820002> 0xfc efuse
:02000004008377
:01000000FF00                                                               // 830000> 0xff lock
:02000004008476
:030000001E940B40                                                           // 840000> 0x1e 0x94 0x0b signature (ATmega168P, ATmega168PA, LGT8F168P)
:02000004008476
:01000300BC40                                                               // 840003> 0xbc calibration
:00000001FF
```

MCUdude commented:
@stefanrueger `-mem` or `/mem` is now working, but I still can't to that with the signature memory:

```
$ ./avrdude -cdryrun -patmega8535 -UALL,/eeprom:w:/Users/hans/Downloads/m8535_backup.hex
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9308 (probably m8535)

avrdude: processing -U ALL,/eeprom:w:/Users/hans/Downloads/m8535_backup.hex:i
avrdude: reading 8714 bytes for multiple memories from input file /Users/hans/Downloads/m8535_backup.hex
avrdude: 8192 bytes flash in 1 section [0, 0x1fff]: 128 pages and 0 pad bytes
         writing 8192 bytes to flash ...
         writing | ################################################## | 100% 0.00 s 
         reading | ################################################## | 100% 0.00 s 
         8192 bytes of flash verified
avrdude: 1 byte lfuse in 1 section [0, 0]
         writing 1 byte to lfuse (0xbf), 1 byte written, 1 verified
avrdude: 1 byte hfuse in 1 section [0, 0]
         writing 1 byte to hfuse (0xc2), 1 byte written, 1 verified
avrdude: 1 byte lock in 1 section [0, 0]
         writing 1 byte to lock (0xff), 1 byte written, 1 verified

avrdude done.  Thank you.

$ ./avrdude -cdryrun -patmega8515 -UALL,-signature:w:/Users/hans/Downloads/m8535_backup.hex
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9306 (probably m8515)

avrdude: processing -U ALL,-signature:w:/Users/hans/Downloads/m8535_backup.hex:i
avrdude: reading 8714 bytes for multiple memories from input file /Users/hans/Downloads/m8535_backup.hex
avrdude error: signature of ATmega8515 does not match file (ATmega8535)
        use -F to override this check

avrdude done.  Thank you.

```

MCUdude commented:
Here is the output when writing a multimem file to the ATmega4809, when specifying the fuses in the program itself, like so. It took me a while to figure out how the struct should look like, due to the reserved memories in there, so we should mention this in the docs.

``` c
FUSES = 
{
    0x01, // WDTCFG
    0x02, // BODCFG
    0x03, // OSCCFG
    0xFF,
    0xFF,
    0x06, // SYSCFG0
    0x07, // SYSCFG1
    0x08, // APPEND
    0x09, // BOOTEND
};
```

```
$ ./avrdude -cdryrun -patmega4809 -UALL:w:/var/folders/50/f9myvbw51nbgddp3cc57tzp40000gn/T/arduino_build_887794/Blink.ino.hex
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9651 (probably m4809)

avrdude: processing -U ALL:w:/var/folders/50/f9myvbw51nbgddp3cc57tzp40000gn/T/arduino_build_887794/Blink.ino.hex:i
avrdude: reading 761 bytes for multiple memories from input file /var/folders/50/f9myvbw51nbgddp3cc57tzp40000gn/T/arduino_build_887794/Blink.ino.hex
avrdude warning: /var/folders/50/f9myvbw51nbgddp3cc57tzp40000gn/T/arduino_build_887794/Blink.ino.hex has no data for eeprom, skipping ...
avrdude: 752 bytes flash in 1 section [0, 0x2ef]: 6 pages and 16 pad bytes
         writing 752 bytes to flash ...
         writing | ################################################## | 100% 0.00 s 
         reading | ################################################## | 100% 0.00 s 
         752 bytes of flash verified
avrdude: 9 bytes fuses in 1 section [0, 8]
         writing 9 bytes to fuses ..., 9 bytes written, 9 verified
avrdude: 1 byte fuse0/wdtcfg in 1 section [0, 0]
         writing 1 byte to fuse0/wdtcfg (0x01), 1 byte written, 1 verified
avrdude: 1 byte fuse1/bodcfg in 1 section [0, 0]
         writing 1 byte to fuse1/bodcfg (0x02), 1 byte written, 1 verified
avrdude: 1 byte fuse2/osccfg in 1 section [0, 0]
         writing 1 byte to fuse2/osccfg (0x03), 1 byte written, 1 verified
avrdude: 1 byte fuse5/syscfg0 in 1 section [0, 0]
         writing 1 byte to fuse5/syscfg0 (0x06), 1 byte written, 1 verified
avrdude: 1 byte fuse6/syscfg1 in 1 section [0, 0]
         writing 1 byte to fuse6/syscfg1 (0x07), 1 byte written, 1 verified
avrdude: 1 byte fuse7/append in 1 section [0, 0]
         writing 1 byte to fuse7/append (0x08), 1 byte written, 1 verified
avrdude: 1 byte fuse8/bootend in 1 section [0, 0]
         writing 1 byte to fuse8/bootend (0x09), 1 byte written, 1 verified
avrdude warning: /var/folders/50/f9myvbw51nbgddp3cc57tzp40000gn/T/arduino_build_887794/Blink.ino.hex has no data for lock, skipping ...
avrdude warning: /var/folders/50/f9myvbw51nbgddp3cc57tzp40000gn/T/arduino_build_887794/Blink.ino.hex has no data for userrow/usersig, skipping ...

avrdude done.  Thank you.
```

Is it necessary to mention all the memories the file didn't contain, with the full filename and everything? The hex file was compiled using Arduino IDE, and I feel the output is a bit verbose.

How about something like this? Way less text, but no critical information have been left out

```
$ ./avrdude -cdryrun -patmega4809 -UALL:w:/var/folders/50/f9myvbw51nbgddp3cc57tzp40000gn/T/arduino_build_887794/Blink.ino.hex
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9651 (probably m4809)

avrdude: processing -U ALL:w:/var/folders/50/f9myvbw51nbgddp3cc57tzp40000gn/T/arduino_build_887794/Blink.ino.hex:i
avrdude: reading 761 bytes for multiple memories from input file
avrdude: input file has no data for eeprom, skipping ...
avrdude: 752 bytes flash in 1 section [0, 0x2ef]: 6 pages and 16 pad bytes
         writing 752 bytes to flash ...
         writing | ################################################## | 100% 0.00 s 
         reading | ################################################## | 100% 0.00 s 
         752 bytes of flash verified
avrdude: 9 bytes fuses in 1 section [0, 8]
         writing 9 bytes to fuses ..., 9 bytes written, 9 verified
avrdude: 1 byte fuse0/wdtcfg in 1 section [0, 0]
         writing 1 byte to fuse0/wdtcfg (0x01), 1 byte written, 1 verified
avrdude: 1 byte fuse1/bodcfg in 1 section [0, 0]
         writing 1 byte to fuse1/bodcfg (0x02), 1 byte written, 1 verified
avrdude: 1 byte fuse2/osccfg in 1 section [0, 0]
         writing 1 byte to fuse2/osccfg (0x03), 1 byte written, 1 verified
avrdude: 1 byte fuse5/syscfg0 in 1 section [0, 0]
         writing 1 byte to fuse5/syscfg0 (0x06), 1 byte written, 1 verified
avrdude: 1 byte fuse6/syscfg1 in 1 section [0, 0]
         writing 1 byte to fuse6/syscfg1 (0x07), 1 byte written, 1 verified
avrdude: 1 byte fuse7/append in 1 section [0, 0]
         writing 1 byte to fuse7/append (0x08), 1 byte written, 1 verified
avrdude: 1 byte fuse8/bootend in 1 section [0, 0]
         writing 1 byte to fuse8/bootend (0x09), 1 byte written, 1 verified
avrdude: input file has no data for lock, skipping ...
avrdude: input file has no data for userrow/usersig, skipping ...

avrdude done.  Thank you.
```

stefanrueger commented:
@MCUdude @dl8dtl @mcuee Thanks all round for tests and comments.

> `/mem`

I have had second thoughts. `/` is not a good notation for *but not* as in set theory `/set` means `universe \ set`. Now `-mem` and `\mem` (instead of `/mem`) can be used as list-element and they mean *remove* `mem` from list so far, eg, `ALL,-bootrow`

> but I still can't [suppress the signature check by removing from the list] the signature memory

Correct. The signature check is triggered by the presence of a signature in the *file*, and *not* by the presence of the `signature` memory in the memory list. This has the advantage that writing to fuses or any other memory of a part from a multi-memory file (as in `-U fuses:w:mm-file.hex:i`) triggers the sanity check. Otherwise one would need to always add `signature` to the memory list if the sanity check was required. Also, the creator of a multi-memory `.hex` or `srec` file can leave out the signature on file creation to prevent this sanity check, eg, if the file only contains EEPROM and sigrow which might be SW-agnostic.

I have, however, written a function `avr_sig_compatible()` that closely models SW-compatibility of AVR parts and knows about exceptions of parts with different signatures tht *are* SW-compatible: @MCUdude, please review commit https://github.com/avrdudes/avrdude/pull/1828/commits/cfaee38275a121406cf2851f45d78b4559420c7c.

> Is it necessary to mention all the memories the file didn't contain, with the full filename and everything?

Sure, the full pathname is not needed. I have reduced that now to the filename component of the path. I feel that AVRDUDE should *warn* whenever the file didn't contain requested memories: in the example the user explicitly asked for ALL memories to be set/checked, so they should receive a warning if something they asked for is not available. The user can reduce the warnings by only specifying what's in the file, eg, `flash,fuses`. And yes, ALL means all memories of the part, not all memories of the file. It *must* be the former otherwise what would `-U ALL:r:file` mean when the file does not (yet) exist?

Let me know if I have missed some of the comments and/or any problems where this PR breaks something.

And, yes, once this PR is settled, I will (as always) contribute documentation. It would be *great* if @MCUdude could contribute examples how to generate multi-memory `.hex` files by compiling sketches in Arduino (I don't use that environment).

MCUdude commented:
> @MCUdude, please review commit https://github.com/avrdudes/avrdude/commit/cfaee38275a121406cf2851f45d78b4559420c7c.

The commit looks good! AFAIK, it only compares devices that are _truly_ SW compatible. So even though a hex file compiled for the ATmega88 will work on the ATmega168, they aren't back to back compatible, and will fail the test.

> Sure, the full pathname is not needed. I have reduced that now to the filename component of the path. 

Thanks! The output looks suddenly much better. And a warning is fine by me.

> It would be great if @MCUdude could contribute examples how to generate multi-memory .hex files by compiling sketches in Arduino (I don't use that environment).

Sure! I can provide examples for classic AVRs and AVRs with UPDI. Embedding the fuses is pretty straight forward, but I haven't figured out how/if it's possible to embed EEPROM content as well.

stefanrueger commented:
@MCUdude Thanks for checking

> a hex file compiled for the ATmega88 will work on the ATmega168, they aren't back to back compatible, and will fail the test

Yes, this is deliberate. For example, if an application for the ATmega88 makes use of the urclock metadata and urboot `write_page()` function located at `FLASHEND-4+1` there will be tears if that binary is put on an ATmega168 as `FLASHEND` differs in both parts... If a user restores a binary image from an `m88` to an `m168` they really should be affirming with `-F` that they know what they are doing even though more often that not this is likely to be OK.

I have in the meantime found (and fixed) an error: AVRDUDE erase the chip automatically if one of the `-U` options writes to flash; this has now been extended to also check memory lists for writing to a flash memory.

Also added documentation.

Would like to merge soon, unless there are further comments. We can still put examples how to generate multi-memory .hex files by compiling sketches into the documentation later on.

Generally, to generate multi-memory files for distribution I recommend to carefully prepare everything in a chip as it should be and then use `-U all:r:distro.hex`. This allows one to use the `config` terminal command to set/check the fuses.

stefanrueger commented:
## Some dryrun tests
```
$ cat mkcheck 

#!/usr/bin/env bash

# Check using randomly generated files
for i in {1..4}; do
  # Create random contens, backup to srec file
  avrdude -qq -c dryrun -p 64dd28 -xrandom -xseed=$i -U ALL:r:/tmp/backup.srec:s && echo $i Backup srec OK
  # Create same random contens, backup to hex file
  avrdude -qq -c dryrun -p 64dd28 -xrandom -xseed=$i -U ALL:r:/tmp/backup.hex:I && echo $i Backup hex OK
  # Restore from hex file and compare against srec
  avrdude -qq -c dryrun -p 64dd28 -U ALL:w:/tmp/backup.hex:I -U ALL:v:/tmp/backup.srec && echo $i hex vs srec OK
  # Restore from srec file and compare against hex file
  avrdude -qq -c dryrun -p 64dd28 -U ALL:w:/tmp/backup.srec:s -U ALL:v:/tmp/backup.hex:I && echo $i srec vs hex OK
  echo
done

$ ./mkcheck 

1 Backup srec OK
1 Backup hex OK
1 hex vs srec OK
1 srec vs hex OK

2 Backup srec OK
2 Backup hex OK
2 hex vs srec OK
2 srec vs hex OK

3 Backup srec OK
3 Backup hex OK
3 hex vs srec OK
3 srec vs hex OK

4 Backup srec OK
4 Backup hex OK
4 hex vs srec OK
4 srec vs hex OK
```

## Some *real* tests

```
############
# Create a non-trivial random file
#
$avrdude -p 64dd28 -c dryrun -xinit -xseed=7 -UALL:r:64dd28.bak

avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e961b (probably 64dd28)

avrdude: processing -U ALL:r:64dd28.bak:I
         reading multiple memories ...
         eeprom          | ################################################## | 100% 0.00 s 
         flash           | ################################################## | 100% 0.00 s 
         fuses           | ################################################## | 100% 0.00 s 
         fuse0/wdtcfg    | ################################################## | 100% 0.00 s 
         fuse1/bodcfg    | ################################################## | 100% 0.00 s 
         fuse2/osccfg    | ################################################## | 100% 0.00 s 
         fuse5/syscfg0   | ################################################## | 100% 0.00 s 
         fuse6/syscfg1   | ################################################## | 100% 0.00 s 
         fuse7/codesize  | ################################################## | 100% 0.00 s 
         fuse8/bootsize  | ################################################## | 100% 0.00 s 
         lock            | ################################################## | 100% 0.00 s 
         prodsig/sigrow  | ################################################## | 100% 0.00 s 
         signature       | ################################################## | 100% 0.00 s 
         tempsense       | ################################################## | 100% 0.00 s 
         sernum          | ################################################## | 100% 0.00 s 
         userrow/usersig | ################################################## | 100% 0.00 s 
         sib             | ################################################## | 100% 0.01 s 
         writing 66034 bytes to output file 64dd28.bak

avrdude done.  Thank you.


############
# Write that file to a *real* device
#
$ avrdude -p 64dd28 -c serialupdi -UALL:w:64dd28.bak

avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e961b (probably 64dd28)
avrdude: Note: carrying out an erase cycle as flash memory needs programming (-U ALL:w:...)
         specify the -D option to disable this feature
avrdude: erasing chip

avrdude: processing -U ALL:w:64dd28.bak:i
avrdude: reading 66004 bytes for multiple memories from input file 64dd28.bak
avrdude: 256 bytes eeprom in 1 section [0, 0xff]
         writing 256 bytes to eeprom ...
         writing | ################################################## | 100% 10.26 s 
         reading | ################################################## | 100% 1.06 s 
         256 bytes of eeprom verified
avrdude: 65536 bytes flash in 1 section [0, 0xffff]: 128 pages and 0 pad bytes
         writing 65536 bytes to flash ...
         writing | ################################################## | 100% 9.91 s 
         reading | ################################################## | 100% 8.96 s 
         65536 bytes of flash verified
avrdude: 16 bytes fuses in 1 section [0, 15]
         writing 16 bytes to fuses ..., 16 bytes written, 16 verified
avrdude: 1 byte fuse0/wdtcfg in 1 section [0, 0]
         writing 1 byte to fuse0/wdtcfg (0x00), 1 byte written, 1 verified
avrdude: 1 byte fuse1/bodcfg in 1 section [0, 0]
         writing 1 byte to fuse1/bodcfg (0x00), 1 byte written, 1 verified
avrdude: 1 byte fuse2/osccfg in 1 section [0, 0]
         writing 1 byte to fuse2/osccfg (0x00), 1 byte written, 1 verified
avrdude: 1 byte fuse5/syscfg0 in 1 section [0, 0]
         writing 1 byte to fuse5/syscfg0 (0xd0), 1 byte written, 1 verified
avrdude: 1 byte fuse6/syscfg1 in 1 section [0, 0]
         writing 1 byte to fuse6/syscfg1 (0x08), 1 byte written, 1 verified
avrdude: 1 byte fuse7/codesize in 1 section [0, 0]
         writing 1 byte to fuse7/codesize (0x1e), 1 byte written, 1 verified
avrdude: 1 byte fuse8/bootsize in 1 section [0, 0]
         writing 1 byte to fuse8/bootsize (0x15), 1 byte written, 1 verified
avrdude: 4 bytes lock in 1 section [0, 3]
         writing 4 bytes to lock ..., 4 bytes written, 4 verified
avrdude: 32 bytes userrow/usersig in 1 section [0, 0x1f]: 1 page and 0 pad bytes
         writing 32 bytes to userrow/usersig ..., 32 bytes written, 32 verified

avrdude done.  Thank you.


############
# Make a backup from the *real* device
#
$ avrdude -p 64dd28 -c serialupdi -UALL:r:64dd28-real.bak
avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e961b (probably 64dd28)

avrdude: processing -U ALL:r:64dd28-real.bak:I
         reading multiple memories ...
         eeprom          | ################################################## | 100% 1.06 s 
         flash           | ################################################## | 100% 10.64 s 
         fuses           | ################################################## | 100% 0.07 s 
         fuse0/wdtcfg    | ################################################## | 100% 0.00 s 
         fuse1/bodcfg    | ################################################## | 100% 0.00 s 
         fuse2/osccfg    | ################################################## | 100% 0.00 s 
         fuse5/syscfg0   | ################################################## | 100% 0.00 s 
         fuse6/syscfg1   | ################################################## | 100% 0.00 s 
         fuse7/codesize  | ################################################## | 100% 0.00 s 
         fuse8/bootsize  | ################################################## | 100% 0.00 s 
         lock            | ################################################## | 100% 0.02 s 
         prodsig/sigrow  | ################################################## | 100% 0.03 s 
         signature       | ################################################## | 100% 0.02 s 
         tempsense       | ################################################## | 100% 0.02 s 
         sernum          | ################################################## | 100% 0.07 s 
         userrow/usersig | ################################################## | 100% 0.01 s 
         sib             | ################################################## | 100% 0.00 s 
         writing 66034 bytes to output file 64dd28-real.bak

avrdude done.  Thank you.


############
# Compare backup from real part with the backup from dryrun part
# Only sigrow differs (as expected)
#
$ diff 64dd28-real.bak 64dd28.bak 

2078c2078
< :200000001E961BFF00040A0DFFFFFFFFFFFFFFFF4222536151005917011001030000000011 // 840000> ................B"SaQ.Y......... prodsig/sigrow
---
> :200000001E961BFF54545454FFFFFFFFFFFFFFFF5554535251504F4E4D4C4B4A49484746F2 // 840000> ....TTTT........UTSRQPONMLKJIHGF prodsig/sigrow
2080,2081c2080,2081
< :200040007467F400CBAE58FFFFFFDDFEFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF3C // 840040> tgt.K.X...]~....................
< :20006000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFAA55AA55FFFFFFFF9E // 840060> ........................*U*U....
---
> :20004000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFC0 // 840040> ................................
> :20006000FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFA0 // 840060> ................................
2085c2085
< :0400040000040A0DDD                                                         // 840004> 0x00 0x04 0x0a 0x0d tempsense
---
> :0400040054545454A8                                                         // 840004> 0x54 0x54 0x54 0x54 tempsense
2087c2087
< :1000100042225361510059170110010300000000F2                                 // 840010> B"SaQ.Y......... sernum
---
> :100010005554535251504F4E4D4C4B4A4948474608                                 // 840010> UTSRQPONMLKJIHGF sernum
2091c2091
< :201000004156522020202020503A32443A312D334D32202841332E4B5630304B2E30290040 // 841000> AVR     P:2D:1-3M2 (A3.KV00K.0). sib
---
> :20100000535353535353535353535353535353535353535353535353535353535353535370 // 841000> SSSSSSSSSSSSSSSSSSSSSSSSSSSSSSSS sib
```

MCUdude commented:
@stefanrueger you mentioned that you wanted to include a few Arduino examples on how the fuses could be added to the the docs.

Arduino sketch for ATmega4809 to store fuses in the output hex file:
```c
#include <avr/fuse.h>

FUSES = {
  0x00, // WDTCFG
  0x00, // BODCFG
  0x7E, // OSCCFG
  0xFF, // Reserved
  0xFF, // Reserved
  0xF6, // SYSCFG0
  0xFF, // SYSCFG1
  0x00, // APPEND
  0x00, // BOOTEND
};

void setup() {
}

void loop() {
}
```

Arduino sketch for ATmega328P to store fuses in the output hex file:
```c
#include <avr/fuse.h>

FUSES = {
  0x62, // LFUSE
  0xD9, // HFUSE
  0xFF, // EFUSE
};

void setup() {
}

void loop() {
}
```
